### PR TITLE
Skolemization format update

### PR DESCRIPTION
**Pull request by i-am-a-teapot**
Update for TPTP proof output for Skolemization for #843

### CODE REVIEW

**Comment by i-am-a-teapot** [comment]
The TSTP standard has changed to allow multiple skolemizations now, which this new format *should* be compatible with.
I ran some test files which looked fine, however there may be edge cases which I am not aware of (seems like skolemization works slighlty differently for higher order things)

File: Shell/Skolem.cpp [hunk 1]
```diff
@@ -487,11 +493,11 @@ Formula* Skolem::skolemise (Formula* f)
         if (arity > 0) {
           def = new QuantifiedFormula(FORALL, vArgs.list(), def);
         }
-
-        Unit* defUnit = new FormulaUnit(def,NonspecificInference0(UnitInputType::AXIOM,InferenceRule::SKOLEM_SYMBOL_INTRODUCTION));
-        UnitList::push(defUnit,_skolimizingDefinitions);
+        
+        //Unit* defUnit = new FormulaUnit(def,NonspecificInference0(UnitInputType::AXIOM,InferenceRule::SKOLEM_SYMBOL_INTRODUCTION));
+        //UnitList::push(defUnit,_skolimizingDefinitions);
```
Comment (MichaelRawson) [inline Shell/Skolem.cpp]: I think this was the last use of `SKOLEM_SYMBOL_INTRODUCTION`. Maybe it can be removed now?

File: Kernel/InferenceStore.cpp [hunk 2]
```diff
@@ -82,6 +88,15 @@ void InferenceStore::recordIntroducedSymbol(Unit* u, SymbolType st, unsigned num
   pStack->push(SymbolId(st,number));
 }
 
+void InferenceStore::recordIntroducedSkolemSymbol(Unit* u, SymbolType st, unsigned number, unsigned replacedVar, std::unique_ptr<std::vector<unsigned>> inScopeVars)
```
Comment (MichaelRawson) [inline Kernel/InferenceStore.cpp]: Why `unique_ptr<vector<unsigned>>` rather than just `vector<unsigned>`?

File: Kernel/InferenceStore.cpp [hunk 3]
```diff
@@ -82,6 +88,15 @@ void InferenceStore::recordIntroducedSymbol(Unit* u, SymbolType st, unsigned num
...
+{
+  SymbolStack* pStack;
+  _introducedSymbols.getValuePtr(u->number(),pStack);
+  _introducedSymbolReplacedVars.insert(number, replacedVar);
```
Comment (MichaelRawson) [inline Kernel/InferenceStore.cpp]: All of these fields are keyed by an `unsigned` and global to Vampire, right? What prevents us from Skolemnising `X0` in different formulas and clashing here?

File: Kernel/InferenceStore.cpp [hunk 4]
```diff
@@ -555,19 +570,24 @@ struct InferenceStore::TPTPProofPrinter
...
+
+  std::string getSymbolName(SymbolId sym) {
+    if (sym.first == SymbolType::FUNC ) {
+      return env.signature->functionName(sym.second);
+    } else if (sym.first == SymbolType::PRED){
+      return env.signature->predicateName(sym.second);
+    } else {
+      return env.signature->typeConName(sym.second);
+    }
+  }
```
Comment (MichaelRawson) [inline Kernel/InferenceStore.cpp:579]: Good idea. I feel like this could be promoted to `Signature.hpp` if it doesn't exist already?

File: Shell/Skolem.hpp [hunk 5]
```diff
@@ -103,7 +103,9 @@ class Skolem
   DHMap<unsigned,TermList> _varSorts;
 
   // for some heuristic evaluations after we are done
-  Stack<std::pair<bool, unsigned>> _introducedSkolemSyms;
+  
+  // Type, numberOfSymbol, Variable that was replaced by the skolem symbol, Variables that are in scope for a skolem symbol (for proof recording purposes)
+  Stack<std::tuple<bool, unsigned, unsigned, std::unique_ptr<std::vector<unsigned>>>> _introducedSkolemSyms;
```
Comment (MichaelRawson) [inline Shell/Skolem.hpp]: The first `bool` is whether the Skolem is for sorts or terms, I think. So: could this be a pair `<unsigned, Term *>` where the lhs is the variable being eliminated and the rhs is the Skolem term?

**Review by MichaelRawson — COMMENTED** [review]
Hi @i-am-a-teapot - thanks for the patch! Got some questions about it but it's looking good, we should have this in for CASC. :)

File: Shell/Skolem.cpp (hunk 1 above)
Comment (i-am-a-teapot) [reply Shell/Skolem.cpp]: I agree, I just commented it out since i was not sure whether it may be good to have it in the future. I think we could remove it!

File: Kernel/InferenceStore.cpp (hunk 3 above)
Comment (i-am-a-teapot) [reply Kernel/InferenceStore.cpp]: You are completely right, I also need to key it by the unit number I am afraid!

**Comment by easychair** [comment]
Did I get it right that we drop a step from the proof (output)?

If we have Skolemization, we should have a rule for it. It is essential for
proof-checking.

Or is it something else?

Andrei

**Comment by i-am-a-teapot** [comment]
Skolemization previously consisted of two steps in our proof output: Introducing the symbol and another step which is the eliminating the existential.
The TSTP format is now slightly changed so we need to be more detailed for skolemization (so we keep recording skolemization!). Pretty much the simplest way of adhering to the required format, is to just do everything in one single step (printing all relevant detail in this single step) instead of splitting it in two.

File: Kernel/InferenceStore.cpp (hunk 2 above)
Comment (i-am-a-teapot) [reply Kernel/InferenceStore.cpp]: Using a term now, this is not necessary anymore.

File: Kernel/InferenceStore.cpp (hunk 3 above)
Comment (i-am-a-teapot) [reply Kernel/InferenceStore.cpp]: I now key by the SymbolId (pair of symbolType and identifier) , so it should be unique.

File: Shell/Skolem.hpp (hunk 5 above)
Comment (i-am-a-teapot) [reply Shell/Skolem.hpp]: I modified it to use what you suggested and I *think*, it should be equivalent to check the term kind (if it is sort or not). (I also checked what is happening in Term and that seems to be ok, but maybe I am missing something)

In some tests I ran this seems to work, let me know if you see a problem with it!

File: Kernel/SubstHelper.hpp [hunk 6]
```diff
@@ -529,10 +529,14 @@ Formula* SubstHelper::applyImpl(Formula* f, Applicator& applicator, bool noShari
...
+      TermList newSort = sort;
+      if(sort.isTerm()){
+        newSort = TermList(applyImpl<ProcessSpecVars>(sort.term(), applicator, noSharing));
+      }
       ASS(binding.isVar());
       unsigned newVar = binding.var();
-      VSList::push({newVar, sort}, newVars);
-      if(newVar!=v) {
+      VSList::push({newVar, newSort}, newVars);
+      if(newVar!=v || newSort!=sort) {
```
Comment (i-am-a-teapot) [inline Kernel/SubstHelper.hpp:536]: In case you are wondering what is happening here:
I tried out a problem (see below), and (default) vampire crashed in tptp proof output mode due to not replacing the skolemized type in the polymorphic container(A), leaving an unbound variable. This should fix it I hope.

```
tff(container_type, type, (
    container: $tType > $tType
)).
tff(contains_pred, type, (
    contains: !> [A: $tType] : ( A * container(A) ) > $o
)).
tff(trigger_axiom, axiom, (
    ! [A: $tType] : 
      ! [X: A] : 
        ? [C: container(A)] : contains(A, X, C)
)).

% 4. Conjecture to complete valid TPTP problem state
tff(dummy_conjecture, conjecture, (
    ! [A: $tType] : 
      ! [X: A] : 
        ? [C: container(A)] : contains(A, X, C)
)).
```

**Review by MichaelRawson — APPROVED** [review]

**Comment by MichaelRawson** [comment]
@quickbeam123 - this is now screwing with `SubstHelper` because it wasn't substituting sort variables in quantifier blocks. Could you take a quick look to double check?

File: Shell/Skolem.hpp [hunk 7]
```diff
@@ -103,8 +103,9 @@ class Skolem
   DHMap<unsigned,TermList> _varSorts;
 
   // for some heuristic evaluations after we are done
-  Stack<std::pair<bool, unsigned>> _introducedSkolemSyms;
-  
+
```
Comment (quickbeam123) [inline Shell/Skolem.hpp]: I know this was undocumented before, but since we are here, could we have a comment about what the `bool` and the `Term*` stand for?

**Comment by quickbeam123** [comment]
> @quickbeam123 - this is now screwing with `SubstHelper` because it wasn't substituting sort variables in quantifier blocks. Could you take a quick look to double check?

The change definitely looks like an improvement! Thank you for proposing it, @i-am-a-teapot !

File: Kernel/InferenceStore.cpp (hunk 4 above)
Comment (quickbeam123) [reply Kernel/InferenceStore.cpp:579]: I agree, although `SymbolId` seems to be a local concept of `InferenceStore` at the moment ...

**Review by quickbeam123 — APPROVED** [review]

File: Kernel/InferenceStore.cpp (hunk 4 above)
Comment (i-am-a-teapot) [reply Kernel/InferenceStore.cpp:579]: For now I figured that I would leave it local here (as I believe typically the code knows with what symbol it deals and it is somewhat unique to InferenceStore to need this).
